catalog/migrate/migrate.go

```diff
@@ -0,0 +1,333 @@
+		}
+	}
+
+	commitID, err := m.entryCatalog.AddCommitNoLock(ctx, repo.RepositoryID, newCommit)
+	if err != nil {
+		return fmt.Errorf("commit existing meta range (commit %d): %w", commit.CommitID, err)
+	}
+
+	// tag the commit id (backward compatibility and a way to skip commit in multiple runs)
+	err = m.entryCatalog.CreateTag(ctx, repo.RepositoryID, tagID, commitID)
+	if err != nil {
+		return fmt.Errorf("create tag %s (branch %s, commit %d): %w", mvccRef, branchID, commit.CommitID, err)
+	}
+
+	m.lastCommit = commitID
+	m.branches[branchID] = commitID
+	m.tags[tagID] = commitID
+	return nil
+}
+
+func (m *Migrate) migrateCommits(ctx context.Context, repo *graveler.RepositoryRecord) error {
+	// initialize per repository state
+	m.branches = make(map[graveler.BranchID]graveler.CommitID)
+	m.tags = make(map[graveler.TagID]graveler.CommitID)
+	m.lastCommit = ""
+
+	// select all repository commits and import
+	rows, err := m.selectRepoCommits(ctx, repo)
+	if err != nil {
+		return fmt.Errorf("select commits: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var commit commitRecord
+		err := commit.Scan(rows)
+		if err != nil {
+			return fmt.Errorf("scan commit record: %w", err)
+		}
+		fmt.Printf("Commit %d (%s %d) Author: %s, Date: %s\n",
+			commit.CommitID, commit.BranchName, commit.BranchID, commit.Committer, commit.CreationDate)
+		if err := m.importCommit(ctx, repo, commit); err != nil {
+			return fmt.Errorf("import commit %d: %w", commit.CommitID, err)
+		}
+	}
+	if err := rows.Err(); err != nil {
+		return fmt.Errorf("read commits: %w", err)
+	}
+
+	// branches - update latest commit
+	for branchID, commitID := range m.branches {
+		_, err := m.entryCatalog.UpdateBranch(ctx, repo.RepositoryID, branchID, commitID.Ref())
+		if err != nil {
+			return fmt.Errorf("update branch %s ref %s: %w", branchID, commitID, err)
+		}
+	}
+	return nil
+}
+
+func (m *Migrate) selectRepoCommits(ctx context.Context, repo *graveler.RepositoryRecord) (pgx.Rows, error) {
+	var mvccRepoID int64
+	err := m.db.GetPrimitive(&mvccRepoID, `SELECT id FROM catalog_repositories WHERE name = $1`, repo.RepositoryID)
+	if err != nil {
+		return nil, fmt.Errorf("select mvcc repository id (%s): %w", repo.RepositoryID, err)
+	}
+	rows, err := m.db.WithContext(ctx).Query(`SELECT c.branch_id, b1.name as branch_name, c.commit_id, 
+			c.previous_commit_id, c.committer, c.message, c.creation_date, c.metadata,
+			c.merge_source_branch, COALESCE(b2.name,'') AS merge_source_branch_name, c.merge_source_commit, c.merge_type
+		FROM catalog_commits c
+		LEFT JOIN catalog_branches b1 ON b1.id=c.branch_id
+		LEFT JOIN catalog_branches b2 ON b2.id=c.merge_source_branch
+		WHERE b1.repository_id = $1
+		ORDER BY commit_id`,
+		mvccRepoID)
+	if err != nil {
+		return nil, fmt.Errorf("select commits (%s): %w", repo.RepositoryID, err)
+	}
+	return rows, nil
+}
+
+func (c *commitRecord) Scan(rows pgx.Row) error {
+	return rows.Scan(c.BranchID, c.BranchName, c.CommitID,
```

Comment:
I don't understand Go, AFAICT this cannot fill in most of the fields on c: how can this scan into e.g. c.CommitID; didn't you just take its _value_?

Reply:
You are right, I make this change in the last sec and I didn't have a test to verify the import.